File: solution/739.daily-temperatures/monotonic-stack.ts
/*
 * @lc app=leetcode.cn id=739 lang=typescript
 *
 * https://leetcode-cn.com/problems/daily-temperatures
 */

// @lc code=start
function dailyTemperatures(temperatures: number[]): number[] {
    const result: number[] = [];
    const stack: number[] = [];
    for (const [i, t] of temperatures.entries()) {
        while (stack.length > 0) {
            const i0 = stack[stack.length - 1];
            const t0 = temperatures[i0];
            if (t <= t0) {
                break;
            }
            stack.pop();
            result[i0] = i - i0;
        }
        stack.push(i);
    }
    while (stack.length > 0) {
        result[stack.pop()!] = 0;
    }
    return result;
}
// @lc code=end

export {};

File: solutions/42.trapping-rain-water/src/monotonic_stack.rs
/*
 * @lc app=leetcode.cn id=42 lang=rust
 *
 * https://leetcode.cn/problems/trapping-rain-water/
 */

pub struct Solution;

// @lc code=start
impl Solution {
    pub fn trap(heights: Vec<i32>) -> i32 {
        let mut stack = Vec::new();
        let mut trapped = 0;
        for (i, &h) in heights.iter().enumerate() {
            while let Some(&j) = stack.last() {
                if h < heights[j] {
                    break;
                }
                stack.pop();
                if let Some(&k) = stack.last() {
                    trapped += (i - k - 1) as i32 * (h.min(heights[k]) - heights[j]);
                }
            }
            stack.push(i);
        }
        trapped
    }
}
// @lc code=end
